packages/flutter/lib/src/widgets/focus_scope.dart

```diff
@@ -1,145 +1,369 @@
+  /// things.
+  final FocusOnKeyCallback onKey;
 
-  /// A list of the [FocusScopeNode]s for each [FocusScope] ancestor of
-  /// the given [BuildContext]. The first element of the list is the
-  /// nearest ancestor's [FocusScopeNode].
+  /// Handler called when the focus changes.
   ///
-  /// The returned list does not include the [FocusManager]'s `rootScope`.
-  /// Only the [FocusScopeNode]s that belong to [FocusScope] widgets are
-  /// returned.
+  /// Called with true if this node gains focus, and false if it loses
+  /// focus.
+  final ValueChanged<bool> onFocusChange;
+
+  /// True if this widget will be selected as the initial focus when no other
+  /// node in its scope is currently focused.
+  final bool autofocus;
+
+  /// An optional focus node to use as the focus node for this [Focus] widget.
+  ///
+  /// If one is not supplied, then one will be allocated and managed by this
+  /// widget.
+  final FocusNode focusNode;
+
+  /// Returns the [focusNode] of the [Focus] that most tightly encloses the given
+  /// [BuildContext].
+  ///
+  /// If this node doesn't have a [Focus] widget ancestor, then the
+  /// [FocusManager.rootScope] is returned.
   ///
   /// The [context] argument must not be null.
-  static List<FocusScopeNode> ancestorsOf(BuildContext context) {
+  static FocusNode of(BuildContext context) {
     assert(context != null);
-    final List<FocusScopeNode> ancestors = <FocusScopeNode>[];
-    while (true) {
-      context = context.ancestorInheritedElementForWidgetOfExactType(_FocusScopeMarker);
-      if (context == null)
-        return ancestors;
-      final _FocusScopeMarker scope = context.widget;
-      ancestors.add(scope.node);
-      context.visitAncestorElements((Element parent) {
-        context = parent;
-        return false;
-      });
-    }
+    final _FocusMarker marker = context.inheritFromWidgetOfExactType(_FocusMarker);
+    return marker?.notifier ?? context.owner.focusManager.rootScope;
+  }
+
+  /// Returns true if the nearest enclosing [Focus] widget's node is focused.
+  ///
+  /// A convenience method to allow build methods to write:
+  /// `Focus.hasFocus(context)` to get whether or not the nearest [Focus] above
+  /// them in the widget hierarchy currently has the keyboard focus.
+  static bool hasFocus(BuildContext context) => Focus.of(context).hasFocus;
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(StringProperty('debugLabel', debugLabel, defaultValue: null));
+    properties.add(FlagProperty('autofocus', value: autofocus, ifTrue: 'AUTOFOCUS', defaultValue: false));
+    properties.add(DiagnosticsProperty<FocusScopeNode>('node', focusNode, defaultValue: null));
   }
 
   @override
-  _FocusScopeState createState() => _FocusScopeState();
+  _FocusState createState() => _FocusState();
 }
 
-class _FocusScopeState extends State<FocusScope> {
+class _FocusState extends State<Focus> {
```

Comment:
metanit: i find this class hard to review because the code is in a bit of a weird order. Generally it's easier to review if the code is in the order that it will execute. In this case I think that would mean:

1. fields
1. initState
1. initNode
1. createNode
1. didChangeDependencies
1. didUpdateWidget
1. handleFocusChanged
1. build (which is always last for consistency)